README.md
# PixelMaestro Studio

![latest release](https://img.shields.io/github/release/8bitbuddhist/pixelmaestro-studio.svg) ![license](https://img.shields.io/github/license/8bitbuddhist/pixelmaestro-studio.svg)

PixelMaestro Studio is a desktop application for controlling LED displays using the [PixelMaestro library](https://github.com/8bitbuddhist/PixelMaestro). Using PixelMaestro Studio, you can create custom LED animations, save and load presets, and control LEDs over USB.

![screenshot](screenshot.png)

## Features

- Design rich, full color animations using interactive real-time visual editing tools
- Create time-based animations and transitions using an intuitive event scheduling system
- Upload your changes in real-time to Arduino and other devices over USB
- Save, load, and share your custom animations and configurations

## Usage

Download the latest version of PixelMaestro Studio from the [Releases](https://github.com/8bitbuddhist/PixelMaestro-Studio/releases) page. The [documentation](https://8bitbuddhist.github.io/PixelMaestro-Studio/) explains how to use PixelMaestro Studio and its various features.

To learn more about the PixelMaestro library itself, [visit the repository](https://github.com/8bitbuddhist/PixelMaestro/) or check out the [documentation](https://8bitbuddhist.github.io/PixelMaestro/).

## Running PixelMaestro Studio

Download the latest version of PixelMaestro Studio from the [releases](https://github.com/8bitbuddhist/PixelMaestro-Studio/releases/) page.

### Windows

Download and run `PixelMaestro_Studio.exe`.

### Linux

Download and run `PixelMaestro_Studio`:

```bash
$ wget https://github.com/8bitbuddhist/PixelMaestro-Studio/releases/download/{version tag}/PixelMaestro_Studio
$ chmod +x PixelMaestro_Studio
$ ./PixelMaestro_Studio
```

## Building PixelMaestro Studio

### Build Requirements
- QT 5.15.2 or higher with the QtSerialPort module
- [PixelMaestro core library](https://github.com/8bitbuddhist/PixelMaestro) (included as a submodule)

### Build Instructions (Linux)
1. [Download and install Qt](https://www.qt.io/download) along with the QtSerialPort module.
2. Clone the GitHub repository to your computer:
	
	- `git clone https://github.com/8bitbuddhist/PixelMaestro-Studio.git`
3. Navigate to the newly created repository folder and use `git submodule` to download the PixelMaestro core library:
	```bash
	$ cd PixelMaestro-Studio
	$ git submodule init
	$ git submodule update
	```
4. Use `qmake` to build the project, or if you have Qt Creator installed, open the `PixelMaestro_Studio.pro` file.
	```bash
	qmake PixelMaestro-Studio.pro && make qmake_all
	```

### Building Static Qt

When creating a static build of Qt 5.12.x, `make` might fail with a "bootstrap-private" error. [See this StackOverflow answer for details](https://forum.qt.io/topic/98501/static-compile-error-qt5-12-0-bootstrap-private/8)

To fix this, install the following libs:

```bash
sudo apt install libxkbcommon-dev
sudo apt install '^libxcb.*-dev' libx11-xcb-dev libglu1-mesa-dev libxrender-dev libxi-dev
```

Then, clone the Qt Git sources and create a shadow build:

```bash
$ git clone https://code.qt.io/qt/qt5.git   // This should also work with Qt sources downloaded from the installer
$ cd qt5
$ mkdir build               // shadow build location
$ mkdir install             // where the static Qt libs will be installed
$ cd build
$ ../configure -static -release -prefix ../install -opensource --confirm-license -skip webengine -nomake tests -nomake examples -nomake tools -system-xcb -recheck-all
$ make -j3                  // -j param specifies the number of cores
$ make install              // Compiled static Qt libs will be placed in the install dir
```

### Building MXE (cross-compiling for Windows)

[MXE](https://mxe.cc) is used to build the PixelMaestro Windows executable in Linux.

```bash
$ git clone https://github.com/mxe/mxe.git
$ cd mxe
$ git checkout 886f50722a943f4a20580facc2deb96d4779394f
$ make qtbase qtserialport
```

### Building for WebAssembly

Qt for WebAssembly is used to generate PixelMaestro for websites. Follow the instructions for installing and setting up [Qt for WebAssembly](https://doc.qt.io/qtcreator/creator-setup-webassembly.html).


```bash

```


### Building Docs

Docs are built using Doxygen and Sphinx. To build docs:

1. Install doxygen
2. Run `pip install -r docsrc/requirements.txt`
3. Run `sphinx-build -b html docsrc/ docs/`

## Credits

Icons based on the [Monochrome icon set by Danny Allen](https://store.kde.org/p/1002558), retrieved from [Openclipart](https://openclipart.org/).


CHANGELOG.md
# Changelog
All notable changes to PixelMaestro Studio will be documented in this file.

The format is loosely based on [Keep a Changelog](http://keepachangelog.com/).

## [v0.70] - In Progress

Based on [PixelMaestro v2.2.0](https://github.com/8bitbuddhist/PixelMaestro/releases/tag/v2.2.0).

### Added
- New option to scale a Section along the x and/or y axes.
- Animation and Canvas timers now display the Frame Rate (frames per second) next to the interval.
- Added the Cue Server, which can read and execute Cues sent over the network. You can enable the Cue Server and set the listening port in the Preferences.
- Added the option to hide confirmation boxes. If you want to display them again, click "Reset confirmation dialogs" in the Preferences.

### Changed

- Animation tab: Renamed “cycles” to “frames”.
- Animation tab: Delay Interval slider will now auto-limit the max value based on the Frame Interval.
- Animation tab: Fixed issue where the Current Frame text box only displayed `0`.
- Fixed a bug where disabling X axis scrolling also disabled setting the X axis offset.

## [v0.60] - 2020-03-05

### Added
- Added support for connecting network devices. Enter the IP address and port number in the format <IP>:<Port> in place of a serial device URI. Leaving the port number blank defaults to port 8077.
- Added wrap toggle to Section tab.
- Added icons to palette, animation, and orientation comboboxes.

### Changed
- GUI changes:
	- Broke out Section and Layer controls to the top of the screen.
	- Converted Section and Layer combo boxes into list boxes.
	- Sections are now displayed in a grid. You can change the width of the grid in the Preferences.
	- Reorganized the Canvas and Show tabs.
	- Added highlights to play/pause buttons when clicked.

## [v0.51] - 2019-11-26

### Added
- Added window icons.

### Changed
- Changed time widgets to display time spans instead of milliseconds.

## [v0.50.2] - 2019-07-26

### Added
- Added checkbox for toggling "Code" column in Cue Preview dialog.
- Add ability to set an Animation's center.

### Changed
- Fixed cursor positioning in SectionDrawingArea.
- Migrated docs to Sphinx.

## [v0.50.1] - 2019-06-30

### Added
- Added support for generating "comet-style" palettes.

### Changed
- Device tab: readded flag when uploading Cues to EEPROM.

## [v0.50] - 2019-06-19

### Important Changes

#### Cue Changes

Due to changes in how Cues are formatted, Cuefiles created in earlier versions can no longer be opened in this version.

### Added
- Added option to mirror Sections across the x and/or y axes.
- Added notification icon for when a background action (e.g. a Show) modifies the Maestro.
- Added notification icon for when a device is connected.
- Added the ability to drag and drop Events in the Show tab's Event Queue.
- Added a dialog for editing Events. Double-click an Event in the Event Queue to change its settings.
- Added option to allow Show Events to trigger live device updates. This allows you to trigger commands on a connected device without having to run the Show on the device.
- Added a play/pause button and cycle selector for Animations.

### Changed
- Enabling Maestro outputs has been moved from the Preferences dialog to the Window menu item.
- Redesigned Device Tab.
- Fixed `MaestroController::save_section_to_datastream()` not saving the Canvas frame count.
- Fixed Device Section Map not updating the correct Cue.
- Fixed Canvases not drawing frames larger than 255x255 pixels.
- Fixed Delay getting disabled when fading is enabled.
- Fixed Canvas actions performed on a SectionDrawingArea not updating the cursor before performing the action.
- Reverted broken duplicate Palette name check.
- Increased minimum window width.
- Bumped Qt version to 5.12.3.

### Removed
- Wave animation mirror option.

## [v0.40.1] - 2018-12-06
### Changed
- Fixed issue with populating Grid Size when switching Sections/Layers.

## [v0.40] - 2018-12-03
### Added
- Main Window
	- Added `File > Open and Merge...`, which loads a Cuefile on top of the current Maestro instead of replacing it.
	- Added `File > Open to Event History...`, which loads a Cuefile into the Event History without affecting the running Maestro.
	- Added ability to resize sections of the interface.
	- Added shortcut keys for Save, New, Open, etc.
	- Added confirmation prompt when there are unsaved changes.
- Animation Editor
	- Added cursor when using the Canvas brush tool.
	- Added duplicate name check to Palette Editor.
	- Added lock icon to tabs when `Lock Maestro` is checked in the Show tab.
	- Added check to prevent segfaults when editing certain animations while `Lock Maestro` is enabled in the Show tab.
	- Added buttons to clear the Event History and Event Queue in the Show tab.
	- Added ability to click on a Section to make it the Active Section.
	- Added "replace" tool for Canvases, which replaces a single color with another color.
	- Added ability to re-select a Palette from the Palette drop-down. This lets you reapply a Palette after making changes.
	- Added refresh button to Device tab, which scans for connected devices.
	- Added ability to map Sections between the PC and a device to "translate" live updates to different Sections.
	- Added Section brightness slider (brightness moved from Maestro-level to Section-level).
- Preferences
	- Added ability to customize the max number of events stored in the Event History.

### Changed
- Main Window
	- Added tooltips to several menu options.
	- Renamed several file menu options to be more clear about their actions.
	- Removed unnecessary Cues being saved to new Cuefiles.
	- Viewport now reflects changes to Maestro brightness.
- Animation Editor
	- Moved Maestro play/pause and lock controls from Show tab to the main window.
	- Split MaestroControlWidget into subwidgets.
	- Show timer now runs continuously instead of only when the Show is running.
	- Renamed `Cue Interpreter` dialog to `Cue Preview`.

src/demo/canvasdemo.cpp
/*
 * PatternDemo.cpp - Demonstrates PixelMaestro's Pattern features.
 */

#include "animation/waveanimation.h"
#include "colorpresets.h"
#include "canvas/animationcanvas.h"
#include "drawingarea/maestrodrawingarea.h"
#include "canvasdemo.h"
#include "canvas/fonts/font5x8.h"
#include "controller/maestrocontroller.h"
#include "core/point.h"

namespace PixelMaestroStudio {
	CanvasDemo::CanvasDemo(QWidget* parent, MaestroController* maestro_controller) : MaestroDrawingArea(parent, maestro_controller) {
		Section* section = maestro_controller_->set_sections(1, Point(80, 80));

		Animation* animation = section->set_animation(AnimationType::Wave);
		animation->set_palette(&ColorPresets::Colorwheel_Palette);
		animation->set_timer(250);
		static_cast<WaveAnimation*>(animation)->set_mirror(true);

		AnimationCanvas* canvas = static_cast<AnimationCanvas*>(section->set_canvas(CanvasType::AnimationCanvas));

		// Start drawing
		// First, box off the Canvas into 4 quadrants
		canvas->draw_line(40, 0, 40, 80);
		canvas->draw_line(0, 40, 80, 40);

		// Draw "PixelMaestro" in the top-left corner
		canvas->draw_line(2, 8, 38, 8);

		Font *font = new Font5x8();
		canvas->draw_text(7, 10, font, (char*)"Pixel", 5);
		canvas->draw_text(2, 20, font, (char*)"Maestro", 7);
		delete font;

		canvas->draw_line(2, 30, 38, 30);

		// Draw a few concentric circles in the top-right corner
		canvas->draw_circle(60, 20, 5, true);
		canvas->draw_circle(60, 20, 10, false);
		canvas->draw_circle(60, 20, 15, false);
		canvas->draw_circle(60, 20, 19, false);

		// Draw triangles in the bottom-left corner
		canvas->draw_triangle(11, 71, 31, 51, 26, 66, true);
		canvas->draw_triangle(9, 69, 14, 54, 29, 49, true);

		canvas->draw_line(0, 80, 40, 40);

		canvas->draw_rect(7, 47, 27, 27, false);

		// Draw a peace symbol in the final quadrant
		canvas->draw_circle(60, 60, 15, false);
		canvas->draw_line(60, 45, 60, 75);

		// Draw v-lines
		canvas->draw_line(50, 50, 60, 60);
		canvas->draw_line(70, 50, 60, 60);
	}
}